app/services/net_ssh_agent_service.rb

```diff
@@ -0,0 +1,89 @@
+require 'net/ssh'
+require 'fileutils'
+
+class NetSshAgentService
+  attr_reader :pid
+  attr_reader :sock
+
+  def initialize(agent_socket, ssh_private_key)
+    FileUtils.mkdir_p(File.dirname(agent_socket))
+    agent_details = `ssh-agent -a #{agent_socket}`
+    @sock         = agent_details.split("=")[1].split(" ")[0].chop
+    @pid          = agent_details.split("=")[2].split(" ")[0].chop
+    IO.popen({"SSH_AUTH_SOCK" => @sock, "SSH_AGENT_PID" => @pid}, ["ssh-add", "-"], :mode => 'w') do |f|
+      f.puts(ssh_private_key)
+      if $?.to_i != 0
+        raise StandardError, "Couldn't add key to agent"
+      end
+    end
+  end
+
+  def ssh_exec!(ssh, command)
+    stdout_data, stderr_data = "", ""
+    exit_code, exit_signal   = nil, nil
+
+    ssh.open_channel do |channel|
+      channel.request_pty do |_ch, success_pty|
+        if success_pty
+          channel.exec(command) do |_, success|
+            raise StandardError, "Command \"#{command}\" was unable to execute" unless success
+
+            channel.on_data do |_, data|
+              stdout_data += data
+            end
+
+            channel.on_extended_data do |_, _, data|
+              stderr_data += data
+            end
+
+            channel.on_request("exit-status") do |_, data|
+              exit_code = data.read_long
+            end
+
+            channel.on_request("exit-signal") do |_, data|
+              exit_signal = data.read_long
+            end
+          end
+        else
+          raise StandardError, "Command \"#{command}\" was unable to execute" unless success
+        end
+      end
+    end
+    ssh.loop
+    {
```

Comment:
Is this passing a block to loop or returning a hash.  I think it might be better to rename `loop` to something else as that is a Ruby keyword which happens to take a block so this code is very confusing to look at.